model/src/exec_env.rs

```diff
@@ -92,59 +93,61 @@ impl<State> ExecutionEnv<State> {
     ///
     /// This can be useful if you want to implement a message as
     /// a method of the state to make it callable from other messages.
-    pub fn split_mut(&mut self) -> (&mut EnvHandler, &mut State) {
+    pub fn split_mut(&mut self) -> (&mut EnvHandler<Env>, &mut State) {
         (&mut self.env_handler, &mut self.state)
     }
 }
 
 /// The actual handler for the environment and for dynamic
 /// allocations and deallocations.
-pub struct EnvHandler {
+pub struct EnvHandler<T> {
     /// The dynamic allocator.
     pub dyn_alloc: CellChunkAlloc,
+    env_marker: PhantomData<T>,
 }
 
-impl AllocateUsing for EnvHandler {
+impl<T> AllocateUsing for EnvHandler<T> {
     unsafe fn allocate_using<A>(alloc: &mut A) -> Self
     where
         A: Allocate,
     {
         Self {
             dyn_alloc: AllocateUsing::allocate_using(alloc),
+            env_marker: PhantomData,
         }
     }
 }
 
-impl Initialize for EnvHandler {
+impl<T> Initialize for EnvHandler<T> {
     type Args = ();
 
     fn initialize(&mut self, _: Self::Args) {
         self.dyn_alloc.initialize(())
     }
 }
 
-impl EnvHandler {
+impl<T: Env> EnvHandler<T> {
     /// Returns the address of the current smart contract.
-    pub fn address(&self) -> env::AccountId {
-        env::address()
+    pub fn address(&self) -> T::AccountId {
+        T::address()
     }
 
     /// Returns the balance of the current smart contract.
-    pub fn balance(&self) -> env::Balance {
-        env::balance()
+    pub fn balance(&self) -> T::Balance {
+        T::balance()
     }
 
     /// Returns the caller address of the current smart contract execution.
-    pub fn caller(&self) -> env::AccountId {
-        env::caller()
+    pub fn caller(&self) -> T::AccountId {
+        T::caller()
     }
 
     /// Returns from the current smart contract execution with the given value.
-    pub unsafe fn r#return<T>(&self, val: T) -> !
+    pub unsafe fn r#return<V>(&self, val: V) -> !
     where
-        T: parity_codec::Encode,
+        V: parity_codec::Encode,
     {
-        env::r#return(val)
+        T::r#return(&val.encode()[..])
```

Comment:
Can't we go for `T::r#return(val)` here, too?

Reply:
It's inlining the impl which used to be in the `env::r#return` function which has been removed: https://github.com/paritytech/ink/pull/108/files#diff-aaf76b409bdca20872661cfad19d09c1L97

Reply:
So it is no longer possible to call this in `core`? So users need `ink_model` to actually make use of this abstraction layer that abstracts between the real environment and the test environment or is it different now due to the refactoring?

Reply:
You're right the test-env abstraction is broken now. I'm currently looking into this.

Reply:
Sorted